### ISSUE TITLE
Don't let Aranym call external tools/scripts

### ISSUE THREAD

**Issue by mikrosk** [aranym/aranym#76]
I've been studying aranym network code for a while and I'm surprised that all targets except Cygwin basically call some script or executable externally. This executable must be owned by `root` with the setuid bit set.

In my opinion this dramatically decreases user experience and troubleshooting, esp. when one wishes to use aranym on multiple operating systems:

- the external tool is never the same, sometimes it's a script, sometimes it's a native executable -- user has to remember which one to use
- if there is something wrong with the tool, the error message is very often quite misleading (like "TunTap(0): ERROR: aratapif failed (code 2). Ethernet disabled!" -- only because of missing `root` ownership)
- sometimes the external tool does all the heavy work (like `aratapif.sh`), sometimes one has to do additional configuration configuration steps anyway (like with `aratapif`) ... especialy Linux's `aratapif` seems totally pointless as it just calls one `ifconfig`, exactly that `ifconfig` which is part of 99% of tutorials floating around the net to be put in the `sudo`ed startup script.
- it greatly complicates using the same configuration file (or creating tutorials) for multiple OSes - `ptp` basically just means "call aratapif(.sh)". No external tool = no need for any `ptp` keyword, just two IPs, Netmask and MAC address, everyone would understand that.
- sometimes it even [doesn't work](https://www.atari-forum.com/viewtopic.php?f=27&t=39385&p=410961#p410961) or leads to pointless investigation even for [more experienced users](https://github.com/aranym/aranym/issues/62#issuecomment-766430693).
- `setuid`'ed tools behave differently on different setups (even on different Linux distributions) -- sometimes they allow to have the tool symlinked, sometimes not. Sometimes `sudo` searches user $PATH, sometimes root $PATH. Again, it adds to the confusion and makes creating tutorials hard.
- it doesn't even make sense to call that tool repeatedly, network has to be set one time, before first aranym usage, not by every start

What I propose is:
- get rid of `bridge`/`ptp` and keep the configuration in aranym config file nice and simple
- for each target supply **one** script which would take one parameter (aranym configuration file) which just calls `ifconfig`, set all that crazy stuff about which user doesn't have to have a clue and be done. Or course, this  would require some sanity checks (`ifconfig` vs `ip`, `tunctl` vs `ip`, ...). The main advantage is that we can tell the user what's wrong -- mising `tun/tap` driver, wrong permissions, ...
- for advanced users the script could provide some other options to set [different LAN setups](https://github.com/aranym/aranym/wiki/manual#linux-host-os-networking-setup)

The tutorial would then be:
- set the IP addresses
- run the script for your OS as root/with `sudo`
- run Aranym as user

IMO this is way more comfortable than current chaotic situation with the countless desperate screams about how Aranym is hard to get working on network.

**Comment by joysfera** [comment]
I'd think you haven't studied the networking code enough yet, if you
don't see the reason for our own "ifconfig" with setuid bit set.

Also, I don't know why you mention "sudo" - it's not meant to be used
in ARAnyM networking at all.

And last but not least - IIRC there are two very different types of
guest networking setup on Linux (proxyarp or bridging) and that's why
we have two different kinds of setup. IIRC the proxyarp was kind of a
hack and I don't know if it's still working well. The bridging, on the
other hand, used to be my preferred way that worked great because it
let my Atari computer be part of my LAN.

BTW, I think we were missing another usual type of guest networking and
that is creating a special network just for the guest and letting the
host NAT the outbound traffic. I feel like this is the default setup on
other similar sandboxes/virtual machines (VirtualBox, QEMU) out there.
We should probably explore that - it might be easiest to set up (most
probably does not need 'ifconfig' at all, I'd guess).

However, I do agree 100% with you that the networking setup would need
major refresh and should be far more user friendly. If I had time I
would start studying the VirtualBox and/or QEMU to see how they do that
and copy as much from them as possible.

Petr


Miro Kropáček píše v St 27. 01. 2021 v 23:15 -0800:

**Comment by th-otto** [comment]
I agree with Petr. Your work on an updated tutorial is really highly appreciated, but there is a reason for having a keywords like ptp/bridge in the config file. For the "bridge" configuration, there is no script called at all on startup, and you actually have the situation you described: the bridge has to be set up by a (user written) script before you run aranym.

As to NAT: at least VirtualBox only works with this, because they use a kernel module. That's IMHO way out of our scope. And IIRC, that also requires a DHCP server built into the application.

**Comment by joysfera** [comment]
Yeah, kernel module is of course impossible for us. The "our ifconfig" would work greatly if linux distros didn't strip the setuid bit, obviously.

As for internal DHCP, does usual Atari setup use DHCP client commonly at all? If yes then we might look into implementing one, there could be some extremely simple ones GPLd out there. This is not the one but just an example what people are capable of :-) https://github.com/infusion/node-dhcp

**Comment by joysfera** [comment]
Another simple DHCP server, this time in a normal programming language:
https://github.com/espressif/esp-idf/blob/master/components/lwip/apps/dhcpserver/dhcpserver.c

Wondering if we could use the lwip to create a cross platform guest networking implementation. Would love to dive into that but will not have the required space-time for that anytime soon, I'm afraid.

**Comment by mikrosk** [comment]
@joysfera:

> I'd think you haven't studied the networking code enough yet, if you don't see the reason for our own "ifconfig" with setuid bit set.

Please read my post carefully. I'm not saying I don't see the reason, I just don't agree with that reason. Everything it does can be done without it in a more transparent way.

> Also, I don't know why you mention "sudo" - it's not meant to be used in ARAnyM networking at all.

Oh sure. So commands (directly quoted from the wiki):

```
$ echo 1 >/proc/sys/net/ipv4/ip_forward
$ echo 1 >/proc/sys/net/ipv4/conf/tap0/proxy_arp
$ arp -Ds 192.168.0.114 eth0 pub
$ route add -host 192.168.0.114 tap0
```

or

```
#/bin/bash
#
# All this should be done as root
#
[..]
```

or 

```
$ modprobe tun
$ tunctl -t tap0 -u <your_loginname>
```

execute just by themselves, magically, no Aranym or networking involved.

> And last but not least - IIRC there are two very different types of guest networking setup on Linux (proxyarp or bridging) and that's why we have two different kinds of setup. 

You are completely missing the point. "ptp" JUST CALLS IFCONFIG, NOTHING ELSE. There is no "different guest networking setup", everything has to be done via additional scripts. Aranym has nothing to do with it.

@th-otto:

> For the "bridge" configuration, there is no script called at all on startup, and you actually have the situation you described: the bridge has to be set up by a (user written) script before you run aranym.

My point exactly! And yet we choose to confuse users with "ptp" which does absolutely nothing more than calling one command which can be called in the setup script (which is in 99% cases needed anyway).

---

Don't get me wrong guys, I appreciate all the work you put into Aranym but I think you're way too detached from common user experience. Setting up aranym networking is super-paintful, nobody really knows how to do that, people can rarely get a qualified answer from a developer and if they do, all they get is YET ANOTHER configuration file + setup script with a lot of magic inside WHICH OF COURSE doesn't work on Windows or Mac OS X.

So my goal is to simplify and unify this as much as possible and removing that aratapif/bpf_helper hacks is the first step.

**Comment by joysfera** [comment]
I *think* (can't recall anything so it's more like a guess) that the 'aratapif' has to be called *after* something happens in ARAnyM - that's why it is not called from a startup script but rather by the running ARAnyM. But it's been so long that I don't remember anything. Maybe it's designed so just to keep the IP configuration (address, gateway, ...) at one place and not scattered across host OS and guest OS. Dunno.

I repeat that you have my full support for reworking the networking in any way that will be more user friendly. I suggest to explore the internal DHCP server implementation in ARAnyM if it allowed for less painfull guest setup (say the FreeMiNT would work out-of-the-box). As for OS setup on the host, well, we need to find some least complicated way for each of the major platforms. If the NATed networking led  to zero required setup on the host then that would be the preferable way, I'd say.

**Comment by th-otto** [comment]
Yes, same to me. I'm not an expert at networking at all, and i'm happy that i got it working. And that's maybe one of the main problems: there are just so many configurations (outside of the scope of aranym) that have to be taken into account. And another problem: if you get it wrong, then it is easy to mess up your network in a way that you can't connect to the outside at all anymore.

So feel free to send any patches if you think you can implement such a thing, but i have neither the time nor the knowledge to do that. And i don't think that this will be an easy task.
